app/controllers/api/service_templates_controller.rb

```diff
@@ -4,5 +4,85 @@ class ServiceTemplatesController < BaseController
+      create_atomic(data).tap(&:save!)
+    rescue => err
+      raise BadRequestError, "Could not create Service Template - #{err}"
+    end
+
+    private
+
+    def create_atomic(data)
+      service_template = ServiceTemplate.new(data.except('request_info'))
+      request_info = data['request_info']
+      if service_template.prov_type.include?('generic')
+        # Configure generic service template
+        add_orchestration_template_vars(service_template, request_info)
+        add_ansible_tower_job_template_vars(service_template, request_info)
+      else
+        # Configure non-generic service template
+        service_template.add_resource(create_service_template_request(request_info))
+      end
+      dialog = Dialog.find(request_info['dialog_id']) if request_info && request_info['dialog_id']
+      set_provision_action(service_template, dialog, request_info)
+      set_retirement_reconfigure_action(service_template, dialog, request_info)
+      service_template
+    end
+
+    # Need to set the request for non-generic Service Template
+    def create_service_template_request(request_data)
+      # data must be symbolized
+      request_params = request_data.symbolize_keys
+      wf = MiqProvisionWorkflow.class_for_source(request_params[:src_vm_id]).new(request_params, @auth_user_obj)
+      raise 'Could not find Provision Workflow class for source VM' unless wf
+      request = wf.make_request(nil, request_params)
+      raise 'Could not create valid request' if request == false || !request.valid?
+      request
+    end
+
+    def set_provision_action(service_template, dialog, request_info)
+      fqname = if request_info && request_info['fqname']
+                 request_info['fqname']
+               else
+                 service_template.class.default_provisioning_entry_point(service_template.service_type)
+               end
+      ra = service_template.resource_actions.build(:action => 'Provision')
+      ra.update_attributes(:dialog => dialog, :fqname => fqname)
+    end
+
+    def set_retirement_reconfigure_action(service_template, dialog, request_info)
+      [
+        {:name => 'Reconfigure', :param_key => 'reconfigure_fqname', :method => 'default_reconfiguration_entry_point'},
+        {:name => 'Retirement', :param_key => 'retire_fqname', :method => 'default_retirement_entry_point'}
+      ].each do |action|
+        ra = service_template.resource_actions.build(:action => action[:name], :dialog => dialog)
+        fqname = if request_info && request_info[action[:param_key]]
+                   request_info[action[:param_key]]
+                 else
+                   service_template.class.send(action[:method])
+                 end
+        ra.update_attributes(:fqname => fqname) if fqname
+      end
+    end
+
+    def add_orchestration_template_vars(service_template, request_info)
+      return unless service_template.kind_of?(ServiceTemplateOrchestration)
```

Comment:
it will never work because the type is always `ServiceTemplate` according to https://github.com/jntullo/manageiq/blob/4bf6307a7f67bb9a5223048fa122607ecc653e6a/app/controllers/api/service_templates_controller.rb#L19